### DISCUSSION TITLE
如何让vsmake生成的工程的支持xmake的c_cpp_properties.json？

### DISCUSSION THREAD

**Discussion by luciouskami** [xmake-io/xmake#3854] [Q&A]
如题，xmake下配置无错误(wasm32)，能顺利编译过，但是生成的工程似乎没有导入(或转换)c_cpp_properties.json，，导致工具链中的一些默认的头文件搜素路径没加进去，然后智能感知就找不到了，比如 string、time.h这样的头文件。visualstudio目前应该是不支持c_cpp_properties.json的，请问有什么解决方案吗？
还有一个问题，我看到在生成的vs工程中_WIN32的宏是打开的，但是wasm32的目标平台是wasm，如果定义了_WIN32就可能会错误的引入WindowsSDK的头文件，这部分应该怎么处理呢？此外，wasm32的工具链是clang,"__attribute__((used)) "在当前生成的工程中被视为非法关键字，这个有办法解决吗？

**Comment by waruqi** [comment]
?? vsxmake 不是生成的 vs 工程么。。跟 c_cpp_properties.json 有什么关系，那个不是 vscode 的配置么？两者完全不相关吧

**Comment by luciouskami** [comment]
> ?? vsxmake 不是生成的 vs 工程么。。跟 c_cpp_properties.json 有什么关系，那个不是 vscode 的配置么？两者完全不相关吧

可能我没表达清楚，我的意思是生成vs工程的时候工具链的搜索路径没包含进去，导致智能感知失效了，c_cpp_properties只是一个举例

**Reply by waruqi** [reply]
提供下可复现的 example 工程，我这里可以

**Reply by luciouskami** [reply]
已通过邮件发送

**Reply by waruqi** [reply]
不要走邮件，走 github 附件

**Reply by luciouskami** [reply]
使用xmake create test
然后执行xmake f -p wasm
xmake project -k vsxmake2022 -m "debug,release" 生成vs2022的工程文件
![1](https://github.com/xmake-io/xmake/assets/10268785/775b3d8a-e7a7-4118-81af-982ac3cba906)

xmake可以编译通过，但是Intellisense坏了

[test.zip](https://github.com/xmake-io/xmake/files/11779203/test.zip)

**Reply by waruqi** [reply]
vsxmake 工程仅用于 msvc 编译。。wasm 以及其他平台的支持不了。wasm 走的 env 绑定，可以直接编译，生成的工程在 vs  里面运行，就脱离了 wasm sdk 的 env ，xmake 还是能编，但 vs 是没法知道 wasm env 的情况的，当然 Intellisense 不行了。。

不是所有头文件，都是加显式搜索路径的，std 这种都是 clang 自动找的。。

**Reply by luciouskami** [reply]
我明白你的意思了，那xmake有提供什么api能在生成工程的时候把搜索路径加进去吗？比如emcc那个就是$(env EMSDK)/upstream/emscripten/cache/sysroot/include
因为生成的流程是走xmake的，跟vs没关系，所以我想只要智能提示不出错就行了。

**Reply by waruqi** [reply]
add_includedirs，但是你显式加 sysroot 的，有可能会干扰编译。。

**Reply by luciouskami** [reply]
我有试过add_includedirs 编译没问题，但是智能提示会被vs的预定义宏干扰，比如_WIN32 这种，然后因为winsdk的搜索路径没指定就会导致智能提示有上千个报错，虽然在一般的msvc项目中可以不定义这个宏，但是我不知道在xmake生成的工程中该如何做，我试过显式取消定义，但是好像无效，还有就是类似 attribute((used)) 这种关键字无效，理论上在vs中指定clang工具集就行，因为我试过在vs中指定clang工具集在显式指定clang的位置就能支持clang的关键字，但是我不知道在xmake中该怎么做
下面的截图是我在其他msvc中使用clang编译的的做法：
![2](https://github.com/xmake-io/xmake/assets/10268785/aacef037-e9a7-4b0e-8868-b6c2ee0b96cb)
![3](https://github.com/xmake-io/xmake/assets/10268785/cefd43c0-2e91-437a-927d-a26560d49e6a)

但是xmake生成的工程中，缺少设置的地方：
![xmake_project](https://github.com/xmake-io/xmake/assets/10268785/af956d7f-2b56-46f4-a96f-904ca6999464)
其中平台工具集只能选择msvc143

**Reply by waruqi** [reply]
你那个是 clang-cl 跟 clang 也是有差异的。。不建议在 vsxmake 中搞这种，这边不做支持，仅支持 msvc 你自己研究下吧。。

**Reply by luciouskami** [reply]
好，其实我感觉只要骗一下vs让他把智能提示支持了就行，编译啥的根本跟它没关系

**Reply by luciouskami** [reply]
貌似xmake支持clang-cl的工具链，测试了一下能生成vs项目，我感觉只要照搬一份到wasm项目就行了，这样起码可以指定VC++的头文件搜素路径

![clang](https://github.com/xmake-io/xmake/assets/10268785/edd10580-d083-4422-adb5-c9c42c81e988)
如图所示，智能提示是正常的，__attribute__((used))也支持了。。感觉只要vs用clang-cl的，编译走xmake就行，反正没干扰
